smallrye jwt returns Response despite exception being caught

### Describe the bug

I have a simple quarkus reactive resteasy server

```java
@Path("/")
@ApplicationScoped
public class Hello {

    @GET
    @Path("hello")
    @RolesAllowed({"roleYouNeed"})
    @Blocking
    public Uni<Response> hello(){
        return Uni.createFrom().item(Response.ok("hello").build());
    }
}
        
```

Following the documentation found here I create several exception mappers to return a response body which elaborates on the status code. This one for example works when you do not provide a JWT token to a request that needs it:

```java
@Provider//register as JAXRS provider
@Priority(1)//override the build-int mapper(which has 5001 prio)
@Slf4j
public class UnauthorizedExceptionMapper implements ExceptionMapper<UnauthorizedException> {
    @Override
    public Response toResponse(UnauthorizedException exception) {
        log.warn("There was an incoming request that did not have the correct Authorization");
        return Response.status(Response.Status.UNAUTHORIZED)
                .entity(new ErrorResponseObject().addErrorsItem(new ErrorResponseObjectErrorsInner()
                        .code(String.valueOf(Response.Status.UNAUTHORIZED.getStatusCode()))
                        .detail("Unauthorized")))
                .build();
    }

}
```

This works as expected on my endpoint that uses quarkus-smallrye-jwt with the annotation `@RolesAllowed({"roleYouNeed"})`

When someone sends a token that is truly worthless however a JWTParse Exception occurs and a different exception is raised, the `AuthenticationFailedException`

To catch this one I created 

```java
@Provider
@Priority(1)
@Slf4j
public class AuthenticationExceptionMapper implements ExceptionMapper<AuthenticationFailedException> {
    @Override
    public Response toResponse(AuthenticationFailedException exception) {
        log.warn("There was an incoming request that did not have the correct Authorization");
        return Response.status(Response.Status.UNAUTHORIZED)
                .entity(new ErrorResponseObject().addErrorsItem(new ErrorResponseObjectErrorsInner()
                        .code(String.valueOf(Response.Status.UNAUTHORIZED.getStatusCode()))
                        .detail("Unauthorized")))
                .build();
    }
}
```

This does catch the exception and my warning is logged: `[com.ene.odp.exc.AuthenticationExceptionMapper] (vert.x-eventloop-thread-0) There was an incoming request that did not have the correct Authorization
`. The problem here is that a response has already been returned and I see the following error occurring:

```
2022-07-25 16:22:03,812 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (vert.x-eventloop-thread-1) Request failed: java.io.IOException: java.io.IOException: Failed to write
        at org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveOutputStream.close(ResteasyReactiveOutputStream.java:272)
        at io.quarkus.resteasy.reactive.jackson.runtime.serialisers.BasicServerJacksonMessageBodyWriter.writeResponse(BasicServerJacksonMessageBodyWriter.java:44)
        at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:207)
        at org.jboss.resteasy.reactive.server.core.serialization.DynamicEntityWriter.write(DynamicEntityWriter.java:104)
        at org.jboss.resteasy.reactive.server.handlers.ResponseWriterHandler.handle(ResponseWriterHandler.java:31)
        at org.jboss.resteasy.reactive.server.handlers.ResponseWriterHandler.handle(ResponseWriterHandler.java:15)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:141)
        at org.jboss.resteasy.reactive.server.vertx.VertxResteasyReactiveRequestContext$1$1.handle(VertxResteasyReactiveRequestContext.java:73)
        at org.jboss.resteasy.reactive.server.vertx.VertxResteasyReactiveRequestContext$1$1.handle(VertxResteasyReactiveRequestContext.java:70)
        at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
        at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:63)
        at io.vertx.core.impl.EventLoopContext.lambda$runOnContext$0(EventLoopContext.java:38)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.io.IOException: Failed to write
        at org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveOutputStream.write(ResteasyReactiveOutputStream.java:109)
        at org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveOutputStream.writeBlocking(ResteasyReactiveOutputStream.java:224)
        at org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveOutputStream.close(ResteasyReactiveOutputStream.java:270)
        ... 18 more
Caused by: java.lang.IllegalStateException: Response has already been written
        at io.vertx.core.http.impl.Http1xServerResponse.checkValid(Http1xServerResponse.java:675)
        at io.vertx.core.http.impl.Http1xServerResponse.writeQueueFull(Http1xServerResponse.java:274)
        at io.quarkus.vertx.http.runtime.filters.AbstractResponseWrapper.writeQueueFull(AbstractResponseWrapper.java:486)
        at org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveOutputStream.awaitWriteable(ResteasyReactiveOutputStream.java:116)
        at org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveOutputStream.write(ResteasyReactiveOutputStream.java:82)
```

Why is the response already being returned while I catch this exception? How can I make sure nothing is returned except what I return from the `AuthenticationExceptionMapper`?

### Expected behavior

When I catch the exception in my AuthenticationExceptionMapper no response is returned except one that I return from the mapper. 

### Actual behavior

The smallrye-jwt extension returns a 401 before I can return an adequate response from my exception mapper

### How to Reproduce?

Create a simple server with an endpoint that uses the smallrye-jwt extension. Create an extension mapper that catches the AuthenticationFailedException and then call the endpoint with some gibberish Authorization: Bearer <Token>. 

### Output of `uname -a` or `ver`

21.0.1 Darwin Kernel Version 21.0.1: Tue Sep 14 20:56:24 PDT 2021; root:xnu-8019.30.61~4/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

OpenJDK 64-Bit Server VM GraalVM CE 22.1.0 (build 17.0.3+7-jvmci-22.1-b06, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.9.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

apache-maven-3.8.4

### Additional information

_No response_

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java
@@ -52,6 +52,8 @@ import io.quarkus.runtime.annotations.Recorder;
 import io.quarkus.security.identity.CurrentIdentityAssociation;
 import io.quarkus.vertx.http.runtime.CurrentVertxRequest;
 import io.quarkus.vertx.http.runtime.HttpBuildTimeConfig;
+import io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder.DefaultAuthFailureHandler;
+import io.quarkus.vertx.http.runtime.security.QuarkusHttpUser;
 import io.vertx.core.Handler;
 import io.vertx.ext.web.RoutingContext;
 
@@ -184,7 +186,20 @@ public class ResteasyReactiveRecorder extends ResteasyReactiveCommonRecorder imp
     public Handler<RoutingContext> handler(RuntimeValue<Deployment> deploymentRuntimeValue) {
         Deployment deployment = deploymentRuntimeValue.getValue();
         RestInitialHandler initialHandler = new RestInitialHandler(deployment);
-        return new ResteasyReactiveVertxHandler(initialHandler);
+
+        // ensure our ex. mappers are called when security exceptions are thrown and proactive auth is disabled
+        final Consumer<RoutingContext> eventCustomizer = new Consumer<>() {
+
+            @Override
+            public void accept(RoutingContext routingContext) {
+                // remove default auth failure handler
+                if (routingContext.get(QuarkusHttpUser.AUTH_FAILURE_HANDLER) instanceof DefaultAuthFailureHandler) {
+                    routingContext.remove(QuarkusHttpUser.AUTH_FAILURE_HANDLER);
+                }
+            }
+        };
+
+        return new ResteasyReactiveVertxHandler(eventCustomizer, initialHandler);
     }
 
     /**

--- a/extensions/smallrye-jwt/deployment/pom.xml
+++ b/extensions/smallrye-jwt/deployment/pom.xml
@@ -48,7 +48,7 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
-            <artifactId>quarkus-resteasy-jsonb-deployment</artifactId>
+            <artifactId>quarkus-resteasy-reactive-jsonb-deployment</artifactId>
             <scope>test</scope>
         </dependency>
         <dependency>

--- a/extensions/smallrye-jwt/deployment/src/test/java/io/quarkus/jwt/test/DisabledProactiveAuthFailedExceptionMappingTest.java
+++ b/extensions/smallrye-jwt/deployment/src/test/java/io/quarkus/jwt/test/DisabledProactiveAuthFailedExceptionMappingTest.java
@@ -1,0 +1,48 @@
+package io.quarkus.jwt.test;
+
+import javax.ws.rs.core.Response;
+
+import org.hamcrest.Matchers;
+import org.jboss.resteasy.reactive.server.ServerExceptionMapper;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.security.AuthenticationFailedException;
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+import io.vertx.ext.web.RoutingContext;
+
+public class DisabledProactiveAuthFailedExceptionMappingTest {
+
+    private static final String CUSTOMIZED_RESPONSE = "AuthenticationFailedException";
+    protected static final Class<?>[] classes = { JsonValuejectionEndpoint.class, TokenUtils.class,
+            AuthFailedExceptionMapper.class };
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(classes)
+                    .addAsResource(new StringAsset("quarkus.http.auth.proactive=false\n"), "application.properties"));
+
+    @Test
+    public void testExMapperCustomizedResponse() {
+        RestAssured
+                .given()
+                .auth().oauth2("absolute-nonsense")
+                .get("/endp/verifyInjectedIssuer").then()
+                .statusCode(401)
+                .body(Matchers.equalTo(CUSTOMIZED_RESPONSE));
+    }
+
+    public static class AuthFailedExceptionMapper {
+
+        @ServerExceptionMapper(value = AuthenticationFailedException.class)
+        public Response handle(RoutingContext routingContext) {
+            return Response
+                    .status(401)
+                    .entity(CUSTOMIZED_RESPONSE).build();
+        }
+
+    }
+}

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java
@@ -72,41 +72,7 @@ public class HttpSecurityRecorder {
                 //register the default auth failure handler
                 //if proactive auth is used this is the only one
                 //if using lazy auth this can be modified downstream, to control authentication behaviour
-                event.put(QuarkusHttpUser.AUTH_FAILURE_HANDLER, new BiConsumer<RoutingContext, Throwable>() {
-                    @Override
-                    public void accept(RoutingContext routingContext, Throwable throwable) {
-                        throwable = extractRootCause(throwable);
-                        //auth failed
-                        if (throwable instanceof AuthenticationFailedException) {
-                            authenticator.sendChallenge(event).subscribe().with(new Consumer<Boolean>() {
-                                @Override
-                                public void accept(Boolean aBoolean) {
-                                    if (!event.response().ended()) {
-                                        event.response().end();
-                                    }
-                                }
-                            }, new Consumer<Throwable>() {
-                                @Override
-                                public void accept(Throwable throwable) {
-                                    event.fail(throwable);
-                                }
-                            });
-                        } else if (throwable instanceof AuthenticationCompletionException) {
-                            log.debug("Authentication has failed, returning HTTP status 401");
-                            event.response().setStatusCode(401);
-                            event.response().end();
-                        } else if (throwable instanceof AuthenticationRedirectException) {
-                            AuthenticationRedirectException redirectEx = (AuthenticationRedirectException) throwable;
-                            event.response().setStatusCode(redirectEx.getCode());
-                            event.response().headers().set(HttpHeaders.LOCATION, redirectEx.getRedirectUri());
-                            event.response().headers().set(HttpHeaders.CACHE_CONTROL, "no-store");
-                            event.response().headers().set("Pragma", "no-cache");
-                            event.response().end();
-                        } else {
-                            event.fail(throwable);
-                        }
-                    }
-                });
+                event.put(QuarkusHttpUser.AUTH_FAILURE_HANDLER, new DefaultAuthFailureHandler());
 
                 if (proactiveAuthentication) {
                     Uni<SecurityIdentity> potentialUser = authenticator.attemptAuthentication(event).memoize().indefinitely();
@@ -208,18 +174,6 @@ public class HttpSecurityRecorder {
                 }
             }
         };
-    }
-
-    private Throwable extractRootCause(Throwable throwable) {
-        while ((throwable instanceof CompletionException && throwable.getCause() != null) ||
-                (throwable instanceof CompositeException)) {
-            if (throwable instanceof CompositeException) {
-                throwable = ((CompositeException) throwable).getCauses().get(0);
-            } else {
-                throwable = throwable.getCause();
-            }
-        }
-        return throwable;
     }
 
     public Handler<RoutingContext> permissionCheckHandler() {
@@ -328,5 +282,61 @@ public class HttpSecurityRecorder {
                 });
             }
         };
+    }
+
+    public static final class DefaultAuthFailureHandler implements BiConsumer<RoutingContext, Throwable> {
+
+        private DefaultAuthFailureHandler() {
+        }
+
+        @Override
+        public void accept(RoutingContext event, Throwable throwable) {
+            throwable = extractRootCause(throwable);
+            //auth failed
+            if (throwable instanceof AuthenticationFailedException) {
+                getAuthenticator(event).sendChallenge(event).subscribe().with(new Consumer<Boolean>() {
+                    @Override
+                    public void accept(Boolean aBoolean) {
+                        if (!event.response().ended()) {
+                            event.response().end();
+                        }
+                    }
+                }, new Consumer<Throwable>() {
+                    @Override
+                    public void accept(Throwable throwable) {
+                        event.fail(throwable);
+                    }
+                });
+            } else if (throwable instanceof AuthenticationCompletionException) {
+                log.debug("Authentication has failed, returning HTTP status 401");
+                event.response().setStatusCode(401);
+                event.response().end();
+            } else if (throwable instanceof AuthenticationRedirectException) {
+                AuthenticationRedirectException redirectEx = (AuthenticationRedirectException) throwable;
+                event.response().setStatusCode(redirectEx.getCode());
+                event.response().headers().set(HttpHeaders.LOCATION, redirectEx.getRedirectUri());
+                event.response().headers().set(HttpHeaders.CACHE_CONTROL, "no-store");
+                event.response().headers().set("Pragma", "no-cache");
+                event.response().end();
+            } else {
+                event.fail(throwable);
+            }
+        }
+
+        private static HttpAuthenticator getAuthenticator(RoutingContext event) {
+            return event.get(HttpAuthenticator.class.getName());
+        }
+
+        private static Throwable extractRootCause(Throwable throwable) {
+            while ((throwable instanceof CompletionException && throwable.getCause() != null) ||
+                    (throwable instanceof CompositeException)) {
+                if (throwable instanceof CompositeException) {
+                    throwable = ((CompositeException) throwable).getCauses().get(0);
+                } else {
+                    throwable = throwable.getCause();
+                }
+            }
+            return throwable;
+        }
     }
 }

--- a/independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/ResteasyReactiveVertxHandler.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/main/java/org/jboss/resteasy/reactive/server/vertx/ResteasyReactiveVertxHandler.java
@@ -1,5 +1,7 @@
 package org.jboss.resteasy.reactive.server.vertx;
 
+import java.util.function.Consumer;
+
 import org.jboss.resteasy.reactive.server.handlers.RestInitialHandler;
 
 import io.vertx.core.Handler;
@@ -8,13 +10,16 @@ import io.vertx.ext.web.RoutingContext;
 public class ResteasyReactiveVertxHandler implements Handler<RoutingContext> {
 
     private final RestInitialHandler handler;
+    private final Consumer<RoutingContext> eventCustomizer;
 
-    public ResteasyReactiveVertxHandler(RestInitialHandler handler) {
+    public ResteasyReactiveVertxHandler(Consumer<RoutingContext> eventCustomizer, RestInitialHandler handler) {
         this.handler = handler;
+        this.eventCustomizer = eventCustomizer;
     }
 
     @Override
     public void handle(RoutingContext event) {
+        eventCustomizer.accept(event);
         handler.beginProcessing(event);
     }
 }

--- a/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/framework/ResteasyReactiveUnitTest.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/framework/ResteasyReactiveUnitTest.java
@@ -396,7 +396,8 @@ public class ResteasyReactiveUnitTest implements BeforeAllCallback, AfterAllCall
                 new VertxRequestContextFactory(), executor);
         fieldInjectionSupport.runtimeInit(testClassLoader, application.getDeployment());
 
-        ResteasyReactiveVertxHandler handler = new ResteasyReactiveVertxHandler(application.getInitialHandler());
+        ResteasyReactiveVertxHandler handler = new ResteasyReactiveVertxHandler(ev -> {
+        }, application.getInitialHandler());
         String path = application.getPath();
         Route route = router.route(path);
         for (Consumer<Route> customizer : routeCustomisers) {

